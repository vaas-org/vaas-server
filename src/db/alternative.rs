use super::{issue::IssueId, DbExecutor};
use crate::{async_message_handler_with_span, span::AsyncSpanHandler};
use actix::prelude::*;
use actix_interop::with_ctx;
use color_eyre::eyre::Report;
use serde::{Deserialize, Serialize};
use sqlx::types::Uuid;
use tracing::{debug, instrument};

#[derive(Clone, Hash, PartialEq, Eq, Debug, Deserialize, Serialize, sqlx::Type)]
#[sqlx(transparent)]
pub struct AlternativeId(pub Uuid);

impl AlternativeId {
    pub fn new() -> Self {
        Self(Uuid::new_v4())
    }
}

#[derive(Clone, Debug)]
pub struct InternalAlternative {
    pub id: AlternativeId,
    pub title: String,
    pub issue_id: IssueId,
}

#[derive(Message, Clone, Debug)]
#[rtype(result = "Result<Vec<InternalAlternative>, Report>")]
pub struct AlternativesForIssueId(pub IssueId);

async_message_handler_with_span! {
    impl AsyncSpanHandler<AlternativesForIssueId> for DbExecutor {
        #[instrument]
        async fn handle(msg: AlternativesForIssueId) -> Result<Vec<InternalAlternative>, Report> {
            let pool = with_ctx(|a: &mut DbExecutor, _| a.pool());
            let alternative_id = msg.0;
            let uuid = alternative_id.0;
            debug!("Retrieving alternative by id {id}", id = uuid);
            let user = sqlx::query_as!(InternalAlternative,
                                r#"
                                SELECT id as "id: _", title, issue_id as "issue_id: _" FROM alternatives WHERE issue_id = $1
                                "#, uuid
                            ).fetch_all(&pool).await?;
            Ok(user)
        }
    }
}
