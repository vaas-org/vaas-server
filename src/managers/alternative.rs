use uuid::Uuid;

#[derive(Clone, Hash, PartialEq, Eq, Debug)]
pub struct AlternativeId(pub String);
impl AlternativeId {
    pub fn new() -> Self {
        Self(Uuid::new_v4().to_hyphenated().to_string())
    }
}
