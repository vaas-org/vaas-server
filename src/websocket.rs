use crate::services;
use crate::services::broadcast::BroadcastActor;
use crate::services::client::ClientActor;
use crate::services::client::{IncomingNewClient, UserId};
use crate::services::vote::{AlternativeId, BroadcastVote, IncomingVoteMessage, VoteActor};
use crate::services::{Login, Service};
use actix::prelude::*;
use actix_web_actors::ws;
use serde::{Deserialize, Serialize};
use slog::{debug, error, info, o, warn};

#[derive(Deserialize)]
struct IncomingLogin {
    user_id: String,
    username: String,
}
#[derive(Deserialize)]
struct IncomingVote {
    alternative_id: String,
    user_id: String, // TODO: should be removed
}
#[derive(Deserialize)]
#[serde(tag = "type")]
enum IncomingMessage {
    #[serde(rename = "vote")]
    Vote(IncomingVote),
    #[serde(rename = "login")]
    Login(IncomingLogin),
}

#[derive(Serialize)]
enum IssueState {
    #[serde(rename = "notstarted")]
    NotStarted,
    #[serde(rename = "inprogress")]
    InProgress,
    #[serde(rename = "finished")]
    Finished,
}

#[derive(Serialize)]
struct Alternative {
    id: String,
    title: String,
}

#[derive(Serialize)]
struct OutgoingVote {
    id: String,
    alternative_id: String,
    user_id: String,
}

#[derive(Serialize)]
struct OutgoingClient {
    id: String,
    username: Option<String>,
}

#[derive(Serialize)]
struct Issue {
    id: String,
    title: String,
    description: String,
    state: IssueState,
    alternatives: Vec<Alternative>,
    votes: Vec<OutgoingVote>,
    max_voters: u32,
    show_distribution: bool,
}

#[derive(Serialize)]
#[serde(tag = "type")]
enum OutgoingMessage {
    #[serde(rename = "issue")]
    Issue(Issue),
    #[serde(rename = "vote")]
    Vote(OutgoingVote),
    #[serde(rename = "client")]
    Client(OutgoingClient),
}

type WsClientId = usize;

pub struct WsClient {
    id: WsClientId,
    service: Addr<Service>,
    logger: slog::Logger,
}

impl WsClient {
    pub fn new(service: Addr<Service>, logger: slog::Logger) -> WsClient {
        let id = 0; // TODO: something smart
        WsClient {
            id,
            service,
            logger: logger.new(o!("id" => id)),
        }
    }
    fn send_json<T: Serialize>(&self, ctx: &mut ws::WebsocketContext<Self>, value: &T) {
        match serde_json::to_string(value) {
            Ok(json) => ctx.text(json),
            Err(err) => error!(
                self.logger,
                "Failed to convert to JSON {error}",
                error = err.to_string()
            ),
        }
    }
}

impl Actor for WsClient {
    type Context = ws::WebsocketContext<Self>;

    fn started(&mut self, ctx: &mut Self::Context) {
        info!(self.logger, "New ws client");
        let addr = ctx.address();
        let connect = services::Connect { addr };
        self.service.do_send(connect.clone());
        BroadcastActor::from_registry().do_send(connect.clone());
        ClientActor::from_registry().do_send(connect);
    }

    fn stopped(&mut self, ctx: &mut Self::Context) {
        info!(self.logger, "Ws client left");
        let addr = ctx.address();
        BroadcastActor::from_registry().do_send(services::Disonnect { addr });
    }
}

// Incoming messages from ws
impl StreamHandler<Result<ws::Message, ws::ProtocolError>> for WsClient {
    fn handle(&mut self, msg: Result<ws::Message, ws::ProtocolError>, _ctx: &mut Self::Context) {
        match msg {
            Ok(message) => match message {
                ws::Message::Text(text) => {
                    let m = serde_json::from_str(&text);
                    if let Err(serde_error) = m {
                        println!("JSON error {:#?}", serde_error);
                        return;
                    }
                    let m = m.unwrap();
                    match m {
                        IncomingMessage::Vote(vote) => {
                            debug!(self.logger, "Incoming vote");
                            let vote_actor = VoteActor::from_registry();
                            let user_id = vote.user_id; // TODO: Should not be sent by client
                            let alternative_id = vote.alternative_id;
                            vote_actor.do_send(IncomingVoteMessage(
                                UserId(user_id),
                                AlternativeId(alternative_id),
                            ));
                        }
                        IncomingMessage::Login(login) => {
                            debug!(
                                self.logger,
                                "Incoming login {} {}", login.user_id, login.username
                            );
                            let user_actor = ClientActor::from_registry();
                            user_actor.do_send(Login {
                                user_id: login.user_id,
                                username: login.username,
                            });
                        }
                    }
                }
                _ => {
                    warn!(self.logger, "Client sent something else than text");
                }
            },
            Err(err) => {
                error!(self.logger, "ProtocolError in StreamHandler {:#?}", err);
            }
        }
    }
}

impl Handler<services::ActiveIssue> for WsClient {
    type Result = ();

    fn handle(&mut self, msg: services::ActiveIssue, ctx: &mut Self::Context) {
        debug!(self.logger, "Handling ActiveIssue event");
        let issue = msg.0;
        self.send_json(
            ctx,
            &OutgoingMessage::Issue(Issue {
                id: issue.id,
                title: issue.title,
                description: issue.description,
                state: match issue.state {
                    services::issue::InternalIssueState::NotStarted => IssueState::NotStarted,
                    services::issue::InternalIssueState::InProgress => IssueState::InProgress,
                    services::issue::InternalIssueState::Finished => IssueState::Finished,
                },
                alternatives: issue
                    .alternatives
                    .into_iter()
                    .map(|alt: services::issue::InternalAlternative| Alternative {
                        id: alt.id,
                        title: alt.title,
                    })
                    .collect(),
                votes: issue
                    .votes
                    .into_iter()
                    .map(|vote: services::vote::InternalVote| OutgoingVote {
                        id: vote.id.0,
                        alternative_id: vote.alternative_id.0,
                        user_id: vote.user_id.0,
                    })
                    .collect(),
                max_voters: issue.max_voters,
                show_distribution: issue.show_distribution,
            }),
        )
    }
}

impl Handler<BroadcastVote> for WsClient {
    type Result = ();

    fn handle(&mut self, msg: BroadcastVote, ctx: &mut Self::Context) {
        let vote = msg.0;
        self.send_json(
            ctx,
            &OutgoingMessage::Vote(OutgoingVote {
                id: vote.id.0,
                alternative_id: vote.alternative_id.0,
                user_id: vote.user_id.0,
            }),
        )
    }
}

impl Handler<IncomingNewClient> for WsClient {
    type Result = ();

    fn handle(&mut self, msg: IncomingNewClient, ctx: &mut Self::Context) {
        let client = msg;

        match client.0.username.clone() {
            Some(u) => {
                debug!(
                    self.logger,
                    "Sending client details back to client {} ({})", client.0.id, u
                );
            }
            None => {
                debug!(
                    self.logger,
                    "Sending client details back to client {}", client.0.id
                );
            }
        };

        self.send_json(
            ctx,
            &OutgoingMessage::Client(OutgoingClient {
                id: client.0.id,
                username: client.0.username,
            }),
        )
    }
}
