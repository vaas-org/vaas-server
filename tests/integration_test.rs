extern crate vaas_server;
use actix_codec::Framed;
use actix_http::ws::Codec;
use actix_web::{test, App};
use actix_web_actors::ws;
use db::alternative::AlternativeId;
use dotenv::dotenv;
use futures::{SinkExt, StreamExt};
use insta::assert_ron_snapshot;
use sqlx::PgPool;
use std::env;
use std::time::Duration;
use tokio::io::{AsyncRead, AsyncWrite};
use tokio::time::timeout;
use vaas_server::{
    db::{self, user::UserId},
    server, websocket,
};
use websocket::{IncomingLogin, IncomingMessage, IncomingReconnect, IncomingVote, OutgoingMessage};

const READ_TIMEOUT_MS: u64 = 200;

macro_rules! frame_message_type {
    ($framed:expr, $message_type:path) => {
        match read_message(&mut $framed)
            .await
            .expect("Unable to read ws frame")
        {
            $message_type(message_type) => message_type,
            _ => panic!("Wrong outgoing message type"),
        }
    };
}

async fn read_message(
    framed: &mut Framed<impl AsyncRead + AsyncWrite, Codec>,
) -> Option<OutgoingMessage> {
    let frame = timeout(Duration::from_millis(READ_TIMEOUT_MS), framed.next()).await;
    // ???
    match frame.ok()??.unwrap() {
        ws::Frame::Text(item) => Some(serde_json::from_slice(&item[..]).unwrap()),
        _ => None,
    }
}

async fn read_messages(
    mut framed: &mut Framed<impl AsyncRead + AsyncWrite, Codec>,
) -> Vec<OutgoingMessage> {
    let mut messages = vec![];
    while let Some(message) = read_message(&mut framed).await {
        messages.push(message);
    }
    messages
}

async fn pg_pool() -> PgPool {
    // TODO: read from some shared config
    dotenv().ok();
    let database_url = env::var("DATABASE_URL").expect("DATABASE_URL not set");
    db::new_pool(&database_url).await.unwrap()
}

#[actix_rt::test]
async fn test_login_user() {
    // Setup test serve
    let pool = pg_pool().await;
    let mut srv = test::start(move || {
        server::register_db_actor(pool.clone());
        server::register_system_actors();
        App::new().configure(|app| server::configure(app))
    });
    let mut framed = srv.ws_at("/ws/").await.unwrap();

    // Send user login
    let message = IncomingMessage::Login(IncomingLogin {
        username: "user".to_owned(),
    });
    let message = serde_json::to_string(&message).unwrap();
    framed.send(ws::Message::Text(message)).await.unwrap();

    let messages = read_messages(&mut framed).await;
    assert_ron_snapshot!(messages, { ".**.id" => "[uuid]" });
}

#[actix_rt::test]
async fn test_reconnect() {
    // Setup test server
    let pool = pg_pool().await;
    let mut srv = test::start(move || {
        server::register_db_actor(pool.clone());
        server::register_system_actors();
        App::new().configure(|app| server::configure(app))
    });
    let mut framed = srv.ws_at("/ws/").await.unwrap();

    // Send user login
    let message = IncomingMessage::Login(IncomingLogin {
        username: "user".to_owned(),
    });
    let message = serde_json::to_string(&message).unwrap();
    framed.send(ws::Message::Text(message)).await.unwrap();

    let messages = read_messages(&mut framed).await;
    assert_ron_snapshot!("reconnect - initial login", messages, { ".**.id" => "[uuid]" });

    // TODO: rewrite to filter?
    let mut session_id = None;
    for message in messages {
        match message {
            OutgoingMessage::Client(client) => {
                session_id = Some(client.id);
                break;
            }
            _ => continue,
        }
    }

    let mut framed = srv.ws_at("/ws/").await.unwrap();
    let message = IncomingMessage::Reconnect(IncomingReconnect {
        session_id: session_id.expect("Session id should exist"),
    });
    let message = serde_json::to_string(&message).unwrap();
    framed.send(ws::Message::Text(message)).await.unwrap();

    let messages = read_messages(&mut framed).await;
    assert_ron_snapshot!("reconnect - reload", messages, { ".**.id" => "[uuid]" });
}

#[actix_rt::test]
async fn test_vote() {
    // Setup test server
    let pool = pg_pool().await;
    let mut srv = test::start(move || {
        server::register_db_actor(pool.clone());
        server::register_system_actors();
        App::new().configure(|app| server::configure(app))
    });
    let mut framed = srv.ws_at("/ws/").await.unwrap();
    let user_id = UserId::new();
    let alternative_id = AlternativeId::new();

    // Send vote
    let message = IncomingMessage::Vote(IncomingVote {
        user_id: user_id.clone(),
        alternative_id: alternative_id.clone(),
    });
    let message = serde_json::to_string(&message).unwrap();
    framed.send(ws::Message::Text(message)).await.unwrap();

    // let client = frame_message_type!(framed, OutgoingMessage::Client);
    // assert_eq!(client.username, None);

    let issue = frame_message_type!(framed, OutgoingMessage::Issue);
    assert_eq!(issue.title, "coronvorus bad??");

    let vote = frame_message_type!(framed, OutgoingMessage::Vote);
    assert_eq!(vote.alternative_id, alternative_id);
    assert_eq!(vote.user_id, user_id);

    // Close connection
    framed
        .send(ws::Message::Close(Some(ws::CloseCode::Normal.into())))
        .await
        .unwrap();

    let item = timeout(Duration::from_millis(READ_TIMEOUT_MS), framed.next())
        .await
        .expect("timeout")
        .unwrap()
        .unwrap();
    assert_eq!(item, ws::Frame::Close(Some(ws::CloseCode::Normal.into())));
}

#[actix_rt::test]
async fn test_connect() {
    // Setup test server
    let pool = pg_pool().await;
    let mut srv = test::start(move || {
        server::register_db_actor(pool.clone());
        server::register_system_actors();
        App::new().configure(|app| server::configure(app))
    });
    let mut framed = srv.ws_at("/ws/").await.unwrap();

    let messages = read_messages(&mut framed).await;

    assert_ron_snapshot!(messages, { ".**.id" => "[uuid]" });
}
