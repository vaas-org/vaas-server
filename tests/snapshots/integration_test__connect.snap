---
source: tests/integration_test.rs
expression: messages
---
[
  Issue(
    type: "issue",
    id: "[uuid]",
    title: "coronvorus bad??",
    description: "yes or yes",
    state: Some(inprogress),
    alternatives: [
      Alternative(
        id: "[uuid]",
        title: "yes",
      ),
      Alternative(
        id: "[uuid]",
        title: "other yes",
      ),
      Alternative(
        id: "[uuid]",
        title: "my name is trump i have control",
      ),
    ],
    votes: Some([]),
    max_voters: Some(10),
    show_distribution: true,
  ),
]
